Code for Within-host diversity improves phylogenetic and transmission reconstruction of SARS-CoV-2 outbreaks
https://www.biorxiv.org/content/10.1101/2022.06.07.495142v1 (not peer-reviewed)

1. Model for within-host diversity
> Sequence simulations using a model with within-host diversity: simulate.R <br/>
> Tree comparison: compareTrees.R<br/>

2. Phylogenetic inference
> Maximum likelihood tree inference: raxml_ng.sh <br/>

3. Whole-genome sequence analysis
> Raw reads mapping: align.sh <br/>
> Variant calling: <br/>

4. Analysis
> Comparison of technical replicates: 1_techReps.R <br/>
> Pairwise comparison of samples: <br/>
> Temporal signal in clusters: <br/>
> Transmission inference: <br/>
